client-js/app/elements/labrad-plot.ts

```diff
@@ -144,164 +105,38 @@ export class Plot extends polymer.Base {
-    if (this.svg) {
-      this.svg.select('#x-label').text(newLabel);
-    }
-  }
-
-  @observe('yLabel')
-  yLabelChanged(newLabel: string, oldLabel: string) {
-    if (this.svg) {
-      this.svg.select('#y-label').text(newLabel);
-    }
-  }
-
-  @observe('mouseMode')
-  mouseModeChanged(newMode: string, oldMode: string) {
-    switch (newMode) {
-      case 'pan':
-        this.$.pan.style.color = 'black';
-        this.$.rect.style.color = '#AAAAAA';
-        break;
-
-      case 'zoomRect':
-        this.$.pan.style.color = '#AAAAAA';
-        this.$.rect.style.color = 'black';
-        break;
-    }
-    this.installMouseListeners();
-  }
-
-  private installMouseListeners() {
-    if (!this.svg) return;
-    switch (this.mouseMode) {
-      case 'pan':
-        this.svg.on('mousedown', null);
-        this.svg.call(this.zoom);
-        this.$.plot.style.cursor = 'auto';
-        this.$.plot.style.cursor = 'all-scroll';
-        break;
-
-      case 'zoomRect':
-        this.svg.on('.zoom', null)
-        this.svg.on('mousedown', () => this.zoomRectangle());
-        this.$.plot.style.cursor = 'auto';
-        this.$.plot.style.cursor = 'crosshair';
-        break;
-    }
-  }
-
-  @observe('drawMode2D')
-  drawMode2DChanged(newMode: string, oldMode: string) {
-    switch (newMode) {
-      case 'dots':
-        this.$.dots.style.color = 'black';
-        this.$.rectfill.style.color = '#AAAAAA';
-        this.$.vargrid.style.color = '#AAAAAA';
-        break;
-
-      case 'rectfill':
-        this.$.dots.style.color = '#AAAAAA';
-        this.$.rectfill.style.color = 'black';
-        this.$.vargrid.style.color = '#AAAAAA';
-        break;
-
-      case 'vargrid':
-        this.$.dots.style.color = '#AAAAAA';
-        this.$.rectfill.style.color = '#AAAAAA';
-        this.$.vargrid.style.color = 'black';
-        break;
-    }
-    if (oldMode && newMode !== oldMode) {
-      setTimeout(() => this.handleZoom(), 0);
-    }
-  }
 
-  @observe('numIndeps')
-  numIndepsChanged(newNum: number, oldNum: number) {
-    this.updatePlotStyles();
+  /**
+   * Redraw the plot and attach resize to the window resize event.
+   * Fires when the component is attached to the DOM.
+   */
+  public attached() {
+    this.redraw_();
+    window.addEventListener('resize', (event) => this.redraw_());
   }
 
-  private updatePlotStyles() {
-    if (!this.svg) return;
-    switch (this.numIndeps) {
-      case 1:
-        this.$$('rect.background').style.fill = '#ffffff';
-        this.$.modes2d.style.visibility = 'hidden';
-        this.is1D = true;
-        this.is2D = false;
-        break;
 
-      case 2:
-        this.$$('rect.background').style.fill = '#222222';
-        this.is1D = false;
-        this.is2D = true;
-        break;
+  /**
+   * Add new data to the plot and re-zoom.
+   * Fires when new data arrives via the socket.
+   */
+  public addData(data: Array<Array<number>>) {
+    if (data.length === 0) return;
```

Comment:
How come we're checking if ```data.length``` has zero length? Does it ever fire if data is empty?